bounded/src/bdd.rs

```diff
@@ -168,21 +167,40 @@ impl Context<'_> {
             valuations[i] = Some(bdd.sat_witness().unwrap());
         }
 
-        println!("found counterexample!");
-        for (i, valuation) in valuations.iter().enumerate() {
-            println!("state {}:", i);
-            let valuation = valuation.clone().unwrap();
-            for relation in self.indices.keys().sorted() {
-                print!("{}: {{", relation);
-                for (elements, (i, _mutable)) in self.indices[relation].iter().sorted() {
-                    if valuation.value(self.vars[*i]) {
-                        print!("{:?}, ", elements);
-                    }
-                }
-                println!("}}");
-            }
-            println!();
-        }
+        // Convert valuations to models
+        let valuations: Vec<_> = valuations.into_iter().map(Option::unwrap).collect();
+        let universe = self
+            .signature
+            .sorts
+            .iter()
+            .map(|s| self.universe[s])
+            .collect();
+        valuations
+            .into_iter()
+            .map(|valuation| {
+                Model::new(
+                    self.signature,
+                    &universe,
+                    self.signature
+                        .relations
+                        .iter()
+                        .map(|r| {
+                            let shape = r
+                                .args
+                                .iter()
+                                .map(|s| cardinality(self.universe, s))
+                                .chain([2])
+                                .collect();
+                            Interpretation::new(&shape, |elements| {
+                                valuation
+                                    .value(self.vars[self.indices[r.name.as_str()][elements].0])
+                                    as usize
+                            })
+                        })
+                        .collect(),
+                )
+            })
+            .collect()
```

Comment:
This function has become quite long. Can you at least factor out this conversion to a separate function (or split further if there's a good way to do so)?